wcfsetup/install/files/lib/http/middleware/Xsrf.class.php

```diff
@@ -34,11 +44,45 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
             self::TOKEN_ATTRIBUTE,
             $xsrfToken
         );
+
+        $hasValidXsrfToken = \hash_equals($xsrfToken, $request->getHeaderLine('x-xsrf-token'));
+
         $request = $request->withAttribute(
             self::HAS_VALID_HEADER_ATTRIBUTE,
-            \hash_equals($xsrfToken, $request->getHeaderLine('x-xsrf-token'))
+            $hasValidXsrfToken
         );
 
+        if (
+            $request->getMethod() !== 'GET'
+            && $request->getMethod() !== 'HEAD'
+            && $this->requestHandler->getActiveRequest()
+        ) {
+            $this->assertHasValidXsrfToken($this->requestHandler->getActiveRequest(), $hasValidXsrfToken);
+        }
+
         return $handler->handle($request);
     }
+
+    private function assertHasValidXsrfToken(Request $request, $hasValidXsrfToken): void
+    {
+        if (!\is_subclass_of($request->getClassName(), RequestHandlerInterface::class)) {
+            // Skip the XSRF check for legacy controllers.
+            return;
+        }
+
+        $reflectionClass = new \ReflectionClass($request->getClassName());
+        if ($reflectionClass->getAttributes('DisableXsrfCheck') !== []) {
+            // Controller has opted out of the XSRF check.
+            return;
+        }
+
+        if (!$hasValidXsrfToken) {
+            throw new InvalidSecurityTokenException();
+        }
+    }
+}
+
+#[\Attribute(\Attribute::TARGET_CLASS)]
+final class DisableXsrfCheck
+{
 }
```

Comment:
This should be placed in a dedicated file/namespace for autoloading / autocompletion to work properly.